### PR TITLE
[Soft Delete] - API updates for soft delete + retrieval

### PR DESCRIPTION
**Pull request by cscharf**
# Soft Delete and Restore Ciphers
Bitwarden currently provides traditional CRUD (Create, Read, Update, and Delete) operations for the storage of Ciphers (items) from a user's vault. Based on community demand and feedback this change implements "soft delete" as a new construct (`PUT /delete`) which perhaps a distinct, auditable/logged, and atomic `UPDATE` operations against the Cipher table to mark it with a deleted date. With the ability to "soft delete" also comes the need to "restore" (`PUT /restore`) soft-deleted items which behaves in the exact same why, however setting the `DeletedDate` field to `null`.

The updates for the server project are as follows:

- Stored Procedures
-- **Create and Updates** - deleted date parameter added but ignored
-- **Delete** - revert back to original delete, hard-delete
-- **Soft Delete** - atomic update, sets Deleted Date and updates Revision Date
-- **Restore** - atomic update, sets Deleted Date to NULL and updates Revision Date
- C#
-- **CipherController** - added new PUT methods for /delete and /restore to match the pattern of DELETE /; reverted DELETE and GET methods that were previously filtering on deleted items (leaving that to the individual clients)
-- **CipherService** - added new SoftDelete and Restore methods for handling those operations. Reverted Get and Delete methods to remove the previously added flag parameters
-- **CipherRepository** - Wrapped new stored procedures and reverted previously modified methods

Tested using cURL.

### CODE REVIEW

File: src/Api/Controllers/CiphersController.cs [hunk 1]
```diff
@@ -94,8 +94,12 @@ public async Task<ListResponseModel<CipherDetailsResponseModel>> Get()
             Dictionary<Guid, IGrouping<Guid, CollectionCipher>> collectionCiphersGroupDict = null;
             if (hasOrgs)
             {
+                var keyMatches = ciphers.Select(c => c.Id); // Soft deletes not filtered in tweener "Get" methods
```
Comment (kspearrin) [inline src/Api/Controllers/CiphersController.cs:97]: I am a little confused about what we're doing here. Also, this should probaby be a HashSet for better perf

File: src/Api/Controllers/CiphersController.cs [hunk 2]
```diff
@@ -360,6 +368,83 @@ public async Task DeleteMany([FromBody]CipherBulkDeleteRequestModel model)
...
+        [HttpPut("{id}/delete")]
+        public async Task PutDelete(string id)
+        {
+            var userId = _userService.GetProperUserId(User).Value;
+            var cipher = await _cipherRepository.GetByIdAsync(new Guid(id), userId);
+            if (cipher == null)
+            {
+                throw new NotFoundException();
+            }
+            await _cipherService.SoftDeleteAsync(cipher, userId);
```
Comment (kspearrin) [inline src/Api/Controllers/CiphersController.cs:380]: Did you consider just using existing update methods? Then you wouldn't need all these extra services, repos, and sprocs.

File: src/Sql/dbo/Stored Procedures/CipherDetails_Create.sql [hunk 3]
```diff
@@ -12,7 +12,8 @@
     @FolderId UNIQUEIDENTIFIER,
     @Favorite BIT,
     @Edit BIT, -- not used
-    @OrganizationUseTotp BIT -- not used
+    @OrganizationUseTotp BIT, -- not used
+    @DeletedDate DATETIME2(7) -- not used
```
[the remark this answers is not in the document]
Comment (cscharf) [reply src/Sql/dbo/Stored Procedures/CipherDetails_Create.sql:16]: I thought I was following convention regarding downstream usage, the `@DeletedDate` parameter is not used by the Create procedure other than to not blow up on the table/parameter mapper that Dapper is using.

File: src/Api/Controllers/CiphersController.cs (hunk 1 above)
Comment (cscharf) [reply src/Api/Controllers/CiphersController.cs:97]: This is intended to filter the collections for the ciphers returned (vs. all of them).. but this was a legacy change from my first commit I missed in reverting so I'll revert this.

File: src/Api/Controllers/CiphersController.cs (hunk 2 above)
Comment (cscharf) [reply src/Api/Controllers/CiphersController.cs:380]: We did consider that, leaving all of the logic to the client, however in that same conversion I'm pretty sure we landed on making soft-delete and restore explicit actions called by the client against the API (similar to deleting or updating today). I can certainly dial back in the DB side of things to just use Update and remove all of the other sprocs and repository methods if you would prefer that.

Comment (kspearrin) [reply src/Api/Controllers/CiphersController.cs:372]: I meant using update services and things that the server already provides. Not doing it on the client. For example here, your soft delete service could just set `RevisionDate = DeletedDate = DateTime.UtcNow` and call the replace responsitory.

File: src/Core/Services/Implementations/CipherService.cs [hunk 4]
```diff
@@ -664,6 +664,90 @@ public async Task DeleteFolderAsync(Folder folder)
...
+            }
+
+            if (cipher.DeletedDate.HasValue)
+            {
+                // Already soft-deleted, we can safely ignore this
+                return;
+            }
+
+            cipher.DeletedDate = DateTime.UtcNow;
+            cipher.RevisionDate = DateTime.UtcNow;
```
Comment (kspearrin) [inline src/Core/Services/Implementations/CipherService.cs:681]: I'd suggest making these exactly the same. Could me hundredths of a ms difference here.

```suggestion
            cipher.DeletedDate = cipher.RevisionDate = DateTime.UtcNow;
```

Comment (cscharf) [reply src/Core/Services/Implementations/CipherService.cs]: Made suggested change, agree

**Pull request merged**